src/main/kotlin/org/elm/ide/intentions/TyFunctionGenerator.kt

```diff
@@ -0,0 +1,117 @@
+    protected val declarations by lazy { root.allDeclarations().toList() }
+    /** Additional encoder functions to generate */
+    protected val funcsByTy = mutableMapOf<Ty, GeneratedFunction>()
+    /** Cache of already existing callable expressions that aren't in [funcsByTy] */
+    protected val callablesByTy = mutableMapOf<Ty, String>()
+    /** Unions that need their variants exposed. */
+    protected val unionsToExpose = mutableSetOf<Ref>()
+
+    /** The name of the module in the current file */
+    protected val moduleName by lazy { file.getModuleDecl()?.name ?: "" }
+
+    /** The name to use for the encoder function for each type (does not include the "encoder" prefix) */
+    // There might be multiple types with the same name in different modules, so add the module
+    // name the function for any type with a conflict that isn't defined in this module
+    protected val funcNames by lazy {
+        declarations.groupBy { it.name }
+                .map { (_, decls) ->
+                    decls.map {
+                        it.toRef() to when {
+                            decls.size == 1 -> it.name
+                            else -> it.module.replace(".", "") + it.name
+                        }
+                    }
+                }.flatten().toMap()
+    }
+
+    /** Qualified names of all imported modules */
+    private val importedModules: Set<String> by lazy {
+        file.findChildrenByClass(ElmImportClause::class.java).mapTo(mutableSetOf()) { it.referenceName }
+    }
+
+    /** Get the module qualifier prefix to add to a name */
+    protected fun qualifierFor(ref: Ref): String {
+        return when (ref.module) {
+            moduleName -> ""
+            // We always fully qualify references to modules that we add imports for
+            !in importedModules -> "${ref.module}."
+            else -> ModuleScope.getQualifierForName(file, ref.module, ref.name) ?: ""
+        }
+    }
+
+    /** Prefix [name], defined in [module], with the necessary qualifier */
+    protected fun qual(module: String, name: String) = "${qualifierFor(Ref(module, name))}$name"
+
+    /** Code to insert after the type annotation */
+    abstract val code: String
+
+    /** Imports to add because they are referenced by generated code. */
+    val imports: List<Candidate> by lazy { calculateImports() }
+
+    protected open fun calculateImports(): List<Candidate> {
+        val visibleTypes = ModuleScope.getVisibleTypes(file).all
+                .mapNotNullTo(mutableSetOf()) { it.name?.let { n -> Ref(it.moduleName, n) } }
+        // Hack in List since GlobalScope.getVisibleTypes can't return it
+        val visibleModules = importedModules + "List"
+        return declarations
+                .filter {
+                    it.module != moduleName &&
+                            (it.toRef() in unionsToExpose || it.module !in visibleModules && it.toRef() !in visibleTypes)
+                }
+                .map {
```

Comment:
I could potentially make this `mapTo(mutableSetOf())` to prevent duplicate candidates, but the import merge takes care of that anyway, so I didn't bother.

Reply:
There must be something wrong in my import merge code because I noticed that every time I run "create decoder" in the same file, it will add `required` to the exposed list. Resulting in situations like this:

```
import Json.Decode.Pipeline exposing (required, required, required)
```

Reply:
as a side note, the 'add import' stuff is pretty gnarly. I originally wrote it in Java for Kamil's plugin and then ported it to Kotlin. It also doesn't help that it's so hard for us to emit Elm code right now. I badly need to look into what it would take to do the formatter integration.

Reply:
Hm. That's a bug in the import merge. I am relying on it being idempotent, so I don't check existing imports when emitting code.